source/postprocess/visualization/darcy_velocity.cc

```diff
@@ -0,0 +1,121 @@
+      {
+        AssertThrow(this->introspection().compositional_name_exists("porosity"),
+                    ExcMessage("The 'darcy velocity' postprocessor requires a compositional "
+                               "field called porosity."));
+
+        const unsigned int porosity_idx = this->introspection().find_composition_type(CompositionalFieldDescription::porosity);
+
+        const double velocity_scaling_factor =
+          this->convert_output_to_years() ? year_in_seconds : 1.0;
+
+        MaterialModel::MaterialModelInputs<dim> in(input_data, this->introspection());
+        MaterialModel::MaterialModelOutputs<dim> out(in.n_evaluation_points(), this->n_compositional_fields());
+        MeltHandler<dim>::create_material_model_outputs(out);
+        this->get_material_model().evaluate(in, out);
+        MaterialModel::MeltOutputs<dim> *fluid_out = out.template get_additional_output<MaterialModel::MeltOutputs<dim>>();
+        AssertThrow(std::isfinite(fluid_out->fluid_viscosities[0]),
+                    ExcMessage("To compute the Darcy velocity the material model needs to provide the melt material model "
+                               "outputs. At least the fluid viscosity was not computed, or is not a number."));
+
+        for (unsigned int q=0; q<in.n_evaluation_points(); ++q)
+          {
+            const double porosity = std::max(in.composition[q][porosity_idx], 1e-10);
+            const Tensor<1,dim> gravity = this->get_gravity_model().gravity_vector(in.position[q]);
+            const double solid_density = out.densities[q];
+            const double fluid_viscosity = fluid_out->fluid_viscosities[q];
+            const double fluid_density = fluid_out->fluid_densities[q];
+            const double permeability = fluid_out->permeabilities[q];
+            const Tensor<1,dim> solid_velocity = in.velocity[q];
+            const Tensor<1,dim> darcy_velocity = solid_velocity -
+                                                 permeability / fluid_viscosity / porosity * gravity *
+                                                 (solid_density - fluid_density) * velocity_scaling_factor;
+
+            for (unsigned int k=0; k<dim; ++k)
+              computed_quantities[q](k) = darcy_velocity[k];
+          }
+      }
+    }
+  }
+}
+
+
+// explicit instantiations
+namespace aspect
+{
+  namespace Postprocess
+  {
+    namespace VisualizationPostprocessors
+    {
+      ASPECT_REGISTER_VISUALIZATION_POSTPROCESSOR(DarcyVelocity,
+                                                  "darcy velocity",
+                                                  "A visualization output object that outputs the Darcy velocity "
+                                                  "vector. This postprocessor requires a compositional field named "
+                                                  "'porosity' which must be advected using the 'darcy field' "
+                                                  "compositional field method. "
```

Comment:
Is this still up-to-date? I think the postprocessor now also works for porosity, if it is not advected with the 'darcy field' compositional field method. But it does require a material model that provides the melt output variables, maybe mention that instead?